File: src/leetCodeProblems/PriorityQueue/MinDeletionsForUniqueString1647.java
package leetCodeProblems.PriorityQueue;

/**
 * LeetCode - https://leetcode.com/problems/minimum-deletions-to-make-character-frequencies-unique/submissions/
 */

import java.util.*;


public class MinDeletionsForUniqueString1647 {
	    
    /**
     * minDeletions
     * 
     * @param s
     * @return
     */
    public int minDeletions(String s) {
        
    	HashMap<String, Integer> charCountMap = new HashMap<String, Integer>();
        
        PriorityQueue<Integer> countsPriorityQueue = new PriorityQueue<Integer>(Collections.reverseOrder());
        
        int minDeletions = 0;
        
        for (int i=0; i < s.length(); i++) {
            
            String characterStr = String.valueOf(s.charAt(i));
            
            if (charCountMap.containsKey(characterStr)) {
				
				int currentCount = charCountMap.get(characterStr);
				currentCount = currentCount+1;
				
				charCountMap.put(characterStr, currentCount);
			}
			else {
				charCountMap.put(characterStr, 1);
			}
        }
        
        for (Integer count: charCountMap.values()) {
        	countsPriorityQueue.add(count);
        }
        
        while (!countsPriorityQueue.isEmpty()) {
        	
        	int lastCount = countsPriorityQueue.remove(); 
        	
        	//System.out.println("last count ->" + lastCount);
        	
        	if (!countsPriorityQueue.isEmpty() && lastCount == countsPriorityQueue.peek()) {
        		//System.out.println("peek ->" + countsPriorityQueue.peek());
        		
        		if (lastCount-1 > 0) {
        			countsPriorityQueue.add(lastCount-1);
        		}
        		
        		minDeletions++;
        	}
        }
        //System.out.println("After -> " + charCountMap);
        
        return minDeletions;
    }
    
    public static void main(String[] args) {
        
    	MinDeletionsForUniqueString1647 obj = new MinDeletionsForUniqueString1647();
        
        //System.out.println(obj.minDeletions("hello")); // expected output = 2
        //System.out.println(obj.minDeletions("aab")); // // expected output = 0
        //System.out.println(obj.minDeletions("aaabbbccdt")); // // expected output = 2
    	//System.out.println(obj.minDeletions("aaabbbcc")); // // expected output = 2
    	System.out.println(obj.minDeletions("bbcebab")); // // expected output = 2
        //System.out.println(obj.minDeletions("ceabaacb")); // // expected output = 2
        
        return;
        
    }

}

File: src/leetCodeProblems/PriorityQueue/MinDeletionsForUniqueString1647.go
package main

/*
- Leetcode - https://leetcode.com/problems/minimum-deletions-to-make-character-frequencies-unique/description/
- Time - O(n)
- Space - O(n)
*/

import (
	"container/heap"
)

type Character struct {
	char      rune
	frequency int
}
type charPQ []Character

func (h charPQ) Len() int {
	return len(h)
}

// Max-heap
func (h charPQ) Less(i int, j int) bool {
	return h[i].frequency > h[j].frequency
}

func (h charPQ) Swap(i int, j int) {
	h[i], h[j] = h[j], h[i]
}

func (h *charPQ) Push(a any) {
	*h = append(*h, a.(Character))
}

func (h *charPQ) Pop() any {
	l := len(*h)
	res := (*h)[l-1]
	*h = (*h)[:l-1]
	return res
}

func minDeletions(s string) int {
	m := make(map[rune]int)

	for _, v := range s {
		m[v]++
	}

	pq := new(charPQ)

	for k, v := range m {
		heap.Push(pq, Character{k, v})
	}

	out := 0

	for len(*pq) > 1 {
		e1, e2 := heap.Pop(pq).(Character), heap.Pop(pq).(Character)

		if e1.frequency != e2.frequency {
			heap.Push(pq, e2)
		} else {
			if e1.frequency > 1 {
				heap.Push(pq, Character{e1.char, e1.frequency - 1})
			}

			heap.Push(pq, e2)
			out++
		}
	}
	return out
}
